gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/util/OrderabilityComparator.java

```diff
@@ -0,0 +1,234 @@
+     * Sort List, Set, Path, and Map element-by-element in the order presented by their natural iterator.
+     */
+    private static final Comparator<Iterable> iteratableComparator = (f, s) -> {
+        final Iterator fi = f.iterator();
+        final Iterator si = s.iterator();
+
+        while (fi.hasNext() && si.hasNext()) {
+            final int i = INSTANCE.compare(fi.next(), si.next());
+            if (i != 0) {
+                return i;
+            }
+        }
+
+        return fi.hasNext() ? 1 : si.hasNext() ? -1 : 0;
+    };
+
+    /**
+     * Sort Map by entry-set.
+     */
+    private static final Comparator<Map> mapComparator =
+            Comparator.comparing(Map::entrySet, iteratableComparator);
+
+    /**
+     * Sort Map.Entry first by key, then by value.
+     */
+    private static final Comparator<Map.Entry> entryComparator =
+            Comparator.<Map.Entry,Object>comparing(Map.Entry::getKey, INSTANCE).thenComparing(Map.Entry::getValue, INSTANCE);
+
+    /**
+     * Sort unknown types first by classname, then by natural toString().
+     */
+    private static final Comparator<Object> unknownTypeComparator =
+            Comparator.comparing(f -> f.getClass().getName()).thenComparing(Object::toString);
+
+    /**
+     * The comparators that will be used for the types defined above in the {@link #types} array. Be sure to keep them
+     * in sync.
+     */
+    private static final Comparator[] comparators = new Comparator[] {
+        naturalOrderComparator,          // Boolean.class,
+        numberComparator,                // Number.class,
+        naturalOrderComparator,          // Date.class,
+        naturalOrderComparator,          // String.class,
+        naturalOrderComparator,          // UUID.class,
+        elementComparator,               // Vertex.class,
+        elementComparator,               // Edge.class,
+        elementComparator,               // VertexProperty.class,
+        propertyComparator,              // Property.class,
+        iteratableComparator,            // Path.class,
+        iteratableComparator,            // Set.class,
+        iteratableComparator,            // List.class,
+        mapComparator,                   // Map.class,
+        entryComparator,                 // Map.Entry.class,
+    };
+
+    /**
+     * Marker priority for unknown types. A large number means they will appear after known types. Use -1 if you want
+     * them to appear first.
+     */
+    private static final int UNKNOWN_TYPE = 0xffff;
+
+    /**
+     * Return the first index of the Class array for which the supplied object is an instance. Returns
+     * {@link #UNKNOWN_TYPE} if not an instance of any of the supplied classes.
+     */
+    private static final int type(final Class[] types, final Object o) {
+        for (int i = 0; i < types.length; i++) {
+            if (types[i].isInstance(o)) {
+                return i;
+            }
+        }
+        return UNKNOWN_TYPE;
```

Comment:
What are the cases where we legitimately expect an unknown type? It should probably be an exception if we are dealing with a type not defined in the semantics.

Reply:
g.inject on a local tinkergraph? lambdas? maybe others. Probably would never happen over a remote connection. Part of Orderability semantics is no exceptions thrown - total universal order. I will update the semantics documentation if it's not in there already.